File: clients/notion.go
package client

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"time"
)

type Notion struct {
	journalPageID string
	token         string
}

type ResultObject struct {
	Object      string `json:"object"`
	ID          string `json:"id"`
	CreatedTime string `json:"created_time"`
}

type TodaysPageResult struct {
	Object  string         `json:"object"`
	Results []ResultObject `json:"results"`
}

func NewNotion() *Notion {
	token := os.Getenv("NOTION_TOKEN")
	journalPageID := os.Getenv("NOTION_JOURNAL_DB")

	return &Notion{journalPageID, token}
}

func doRequestToNotion(method string, token string, url string, body []byte) ([]byte, error) {
	req, err := http.NewRequest(
		method,
		url,
		bytes.NewBuffer(body),
	)
	if err != nil {
		log.Println(err)
		return make([]byte, 0), err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
	req.Header.Set("Notion-Version", "2022-06-28")

	// Make the HTTP request
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		log.Println(err)
		return make([]byte, 0), err
	}
	defer resp.Body.Close()

	body, err = ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Println(err)
		return make([]byte, 0), err
	}

	return body, nil
}

func (n *Notion) GetTodaysPage() (TodaysPageResult, error) {
	today := time.Now().Format("2006-01-02")

	requestBody := map[string]interface{}{
		"filter": map[string]interface{}{
			"property": "Date",
			"date": map[string]interface{}{
				"on_or_after": today,
			},
		},
	}

	jsonBody, err := json.Marshal(requestBody)
	if err != nil {
		log.Println(err)
		return TodaysPageResult{}, err
	}

	url := fmt.Sprintf("https://api.notion.com/v1/databases/%s/query", n.journalPageID)
	body, err := doRequestToNotion("POST", n.token, url, jsonBody)
	if err != nil {
		log.Println(err)
		return TodaysPageResult{}, err
	}

	// Parse the response JSON into a Response struct
	var response TodaysPageResult
	err = json.Unmarshal(body, &response)
	if err != nil {
		fmt.Println("Error unmarshaling response body:", err)
		return TodaysPageResult{}, err
	}
	return response, nil
}

func (n *Notion) AddLinkToTodaysPage(link string, summary string) error {

	todaysPage, err := n.GetTodaysPage()
	if err != nil {
		log.Println(err)
		return err
	}

	var result ResultObject
	if len(todaysPage.Results) == 0 {
		r, err := n.CreatePageForToday([]string{"Ratatoskr"})
		if err != nil {
			log.Println(err)
			return err
		}
		result = r
	} else {
		result = todaysPage.Results[0]
	}

	requestBody := map[string]interface{}{
		"children": []interface{}{
			map[string]interface{}{
				"object": "block",
				"type":   "paragraph",
				"paragraph": map[string]interface{}{
					"rich_text": []interface{}{
						map[string]interface{}{
							"type": "text",
							"text": map[string]interface{}{
								"content": summary,
							},
						},
					},
				},
			},
			map[string]interface{}{
				"object": "block",
				"type":   "bookmark",
				"bookmark": map[string]interface{}{
					"url": link,
				},
			},
		},
	}

	jsonBody, err := json.Marshal(requestBody)
	if err != nil {
		log.Println(err)
		return err
	}

	url := fmt.Sprintf("https://api.notion.com/v1/blocks/%s/children", result.ID)
	body, err := doRequestToNotion("PATCH", n.token, url, jsonBody)
	if err != nil {
		log.Println(err)
		return err
	}

	fmt.Println(string(body))

	return nil
}

func (n *Notion) CreatePageForToday(tags []string) (ResultObject, error) {
	dayOfWeek := time.Now().Format("Monday")
	today := time.Now().Format("2006-01-02")

	tag_options := make([]interface{}, len(tags))
	for i, tag := range tags {
		tag_options[i] = map[string]interface{}{
			"name": tag,
		}
	}

	requestBody := map[string]interface{}{
		"parent": map[string]interface{}{
			"database_id": n.journalPageID,
		},
		"properties": map[string]interface{}{
			"Tags": map[string]interface{}{
				"multi_select": tag_options,
			},
			"Date": map[string]interface{}{
				"date": map[string]interface{}{
					"start": today,
				},
			},
			"Name": map[string]interface{}{
				"title": []interface{}{
					map[string]interface{}{
						"text": map[string]interface{}{
							"content": dayOfWeek,
						},
					},
				},
			},
		},
	}

	jsonBody, err := json.Marshal(requestBody)
	if err != nil {
		log.Println(err)
		return ResultObject{}, err
	}

	url := "https://api.notion.com/v1/pages"
	body, err := doRequestToNotion("POST", n.token, url, jsonBody)
	if err != nil {
		log.Println(err)
		return ResultObject{}, err
	}

	// Parse the response JSON into a Response struct
	var response ResultObject
	err = json.Unmarshal(body, &response)
	if err != nil {
		fmt.Println("Error unmarshaling response body:", err)
		return ResultObject{}, err
	}

	return response, nil
}

File: src/clients/notion.rs
//allow dead code in this file
#![allow(dead_code)]
use chrono::Utc;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::env;

#[derive(Debug, Serialize, Deserialize)]
struct Notion {
    journal_page_id: String,
    token: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
struct ResultObject {
    object: String,
    id: String,
    created_time: String,
    url: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct TodaysPageResult {
    object: String,
    results: Vec<ResultObject>,
}

impl Notion {
    fn new() -> Self {
        let token = env::var("NOTION_TOKEN").expect("NOTION_TOKEN environment variable not set");
        let journal_page_id =
            env::var("NOTION_JOURNAL_DB").expect("NOTION_JOURNAL_DB environment variable not set");

        Notion {
            journal_page_id,
            token,
        }
    }

    async fn do_request_to_notion(
        &self,
        method: &str,
        url: &str,
        body: Option<Vec<u8>>,
    ) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
        let client = Client::new();
        let request_builder = client.request(reqwest::Method::from_bytes(method.as_bytes())?, url);

        let request_builder = match body {
            Some(body) => request_builder.body(body),
            None => request_builder,
        };

        let request_builder = request_builder
            .header("Content-Type", "application/json")
            .header("Authorization", format!("Bearer {}", self.token))
            .header("Notion-Version", "2022-06-28");

        let response = request_builder.send().await?;
        let response_body = response.bytes().await?;

        Ok(response_body.to_vec())
    }

    async fn get_todays_page(&self) -> Result<TodaysPageResult, Box<dyn std::error::Error>> {
        let today = Utc::now().format("%Y-%m-%d").to_string();

        let request_body = json!({
            "filter": {
                "property": "Date",
                "date": {
                    "on_or_after": today
                }
            }
        });

        let json_body = serde_json::to_vec(&request_body)?;
        let url = format!(
            "https://api.notion.com/v1/databases/{}/query",
            self.journal_page_id
        );

        let body = self
            .do_request_to_notion("POST", &url, Some(json_body))
            .await?;
        let response: TodaysPageResult = serde_json::from_slice(&body)?;

        Ok(response)
    }

    async fn add_link_to_todays_page(
        &self,
        link: &str,
        summary: &str,
    ) -> Result<ResultObject, Box<dyn std::error::Error>> {
        let todays_page = self.get_todays_page().await?;

        let result = if todays_page.results.is_empty() {
            self.create_page_for_today(&["Ratatoskr"]).await? // Assuming create_page_for_today function exists
        } else {
            todays_page.results[0].clone()
        };

        let request_body = json!({
            "children": [
                {
                    "object": "block",
                    "type": "paragraph",
                    "paragraph": {
                        "rich_text": [
                            {
                                "type": "text",
                                "text": {
                                    "content": summary
                                }
                            }
                        ]
                    }
                },
                {
                    "object": "block",
                    "type": "bookmark",
                    "bookmark": {
                        "url": link
                    }
                }
            ]
        });

        let json_body = serde_json::to_vec(&request_body)?;
        let url = format!("https://api.notion.com/v1/blocks/{}/children", result.id);

        self.do_request_to_notion("PATCH", &url, Some(json_body))
            .await?;

        Ok(result)
    }

    async fn create_page_for_today(
        &self,
        tags: &[&str],
    ) -> Result<ResultObject, Box<dyn std::error::Error>> {
        let day_of_week = Utc::now().format("%A").to_string();
        let today = Utc::now().format("%Y-%m-%d").to_string();

        let tag_options = tags
            .iter()
            .map(|tag| json!({ "name": tag }))
            .collect::<Vec<_>>();

        let request_body = json!({
            "parent": {
                "database_id": self.journal_page_id
            },
            "properties": {
                "Tags": {
                    "multi_select": tag_options
                },
                "Date": {
                    "date": {
                        "start": today
                    }
                },
                "Name": {
                    "title": [
                        {
                            "text": {
                                "content": day_of_week
                            }
                        }
                    ]
                }
            }
        });

        let json_body = serde_json::to_vec(&request_body)?;
        let url = "https://api.notion.com/v1/pages";

        let body = self
            .do_request_to_notion("POST", url, Some(json_body))
            .await?;
        let response: ResultObject = serde_json::from_slice(&body)?;

        Ok(response)
    }
}
